### PR TITLE
deps: bump `ibc-proto-rs` to v0.38.0 + some other dependency upgrades

### PR DESCRIPTION
**Pull request by Farhad-Shabani**
## Description

- Upgrades ibc-proto-rs to v0.38.0
- Upgrades sp-* dependencies
- Upgrades `type-builder` to v0.18.0
- The `encode_vec` method of the `ConsensusState` trait takes `self` by value to avoid unnecessary cloning 

- Integration tests: https://github.com/informalsystems/basecoin-rs/pull/143
______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests.
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [x] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: .changelog/unreleased/breaking-changes/949-bump-ibc-proto-rs-to-v0380.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+- Bump `ibc-proto-rs` to v0.38.0
+ ([#949](https://github.com/cosmos/ibc-rs/issues/949))
```
Comment (rnbguy) [inline .changelog/unreleased/breaking-changes/949-bump-ibc-proto-rs-to-v0380.md]:
```suggestion
  ([\#949](https://github.com/cosmos/ibc-rs/issues/949))
```

File: ci/no-std-check/Cargo.toml [hunk 2]
```diff
@@ -15,10 +15,10 @@ tendermint = { version = "0.34", default-features = false }
...
 tendermint-light-client-verifier = { version = "0.34", default-features = false, features = ["rust-crypto"] }
 
-sp-core = { version = "24.0.0", default-features = false, optional = true }
-sp-io = { version = "26.0.0", default-features = false, optional = true }
-sp-runtime = { version = "27.0.0", default-features = false, optional = true }
-sp-std = { version = "11.0.0", default-features = false, optional = true }
+sp-core = { version = "25.0.0", default-features = false, optional = true }
+sp-io = { version = "27.0.0", default-features = false, optional = true }
+sp-runtime = { version = "28.0.0", default-features = false, optional = true }
+sp-std = { version = "12.0.0", default-features = false, optional = true }
```
Comment (rnbguy) [inline ci/no-std-check/Cargo.toml:21]: Why are only these dependencies upgraded, but not others?

File: crates/ibc/src/core/ics03_connection/error.rs [hunk 3]
```diff
@@ -17,8 +16,6 @@ pub enum ConnectionError {
     Client(client_error::ClientError),
     /// invalid connection state: expected `{expected}`, actual `{actual}`
     InvalidState { expected: String, actual: String },
-    /// invalid connection end error: `{0}`
-    InvalidConnectionEnd(ProtoError),
```
Comment (rnbguy) [inline crates/ibc/src/core/ics03_connection/error.rs:21]: Why did we have this? Looks like, it's not being used anywhere in the codebase.

**Review by rnbguy — CHANGES_REQUESTED** [review]
@Farhad-Shabani Thanks for taking care of this. :pray: Just requested a change.

File: crates/ibc/src/core/ics03_connection/error.rs (hunk 3 above)
Comment (Farhad-Shabani) [reply crates/ibc/src/core/ics03_connection/error.rs:21]: That's right, it's a stale variant. Not being used anymore.

File: ci/no-std-check/Cargo.toml (hunk 2 above)
Comment (Farhad-Shabani) [reply ci/no-std-check/Cargo.toml:21]: [eff17e4](https://github.com/cosmos/ibc-rs/pull/949/commits/eff17e48da37dfad29e2da9f8993368510badc88)

**Review by rnbguy — APPROVED** [review]
Looks good @Farhad-Shabani :handshake:

**Pull request merged**